### DISCUSSION TITLE
Lesson 7: Incompatible Solidity Version

### DISCUSSION THREAD

**Discussion by danmas95** [smartcontractkit/full-blockchain-solidity-course-py#472] [Q&A] [answered]
In the "Smart Lottery" Project I'm having trouble with the `LinkToken.sol` smart contract. The file should be compiled with `pragma solidity ^0.4.11;` but a compile error is  rised by Brownie. The error is `IncompatibleSolcVersion: Brownie only supports Solidity versions >=0.4.22`.

On github repos ([chainlink-mix](https://github.com/smartcontractkit/chainlink-mix/blob/master/contracts/test/LinkToken.sol) and [chainlink](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.4/LinkToken.sol)) there is only the `v0.4` version available of `LinkToken.sol` file.

I tried to manually add all the files necessary to compile that specific smart contract but it's not a good practise and I had lot's of troubles to migrate everything to `solidity ^0.6.0` (the version used in the course project).

Can you suggest me a solution?

Thanks.

**Comment by cromewar** [comment, marked as answer]
Hello @danmas95 
That's strange as brownie 1.17.1 actually compiles that contract without problem, you could try to can downgrade your brownie version to one compatible with `0.4.11;` like 1.16... 
But if you want a simpler solution just use LinkToken on version 0.6.0 which could be found [here](https://github.com/smartcontractkit/LinkToken/blob/master/contracts/v0.6/LinkToken.sol). Just replace the contract with this one.

Best regards, Cromewar.

**Reply by danmas95** [reply]
My Brownie version is: `Brownie v1.16.4 - Python development framework for Ethereum`. 

Anyway, the "simpler" solution looks interesting but I'm facing another problem: I would like to import necessary file via `brownie-config.yaml` but `smartcontractkit/linktoken` repo does not have a release version (`- smartcontractkit/linktoken@?.?.?.`)
How can I solve this? 

Can I avoid to add every single needed file in my project?

Thanks!

**Reply by cromewar** [reply]
It should work without the `@?-?.?` as it would take the latest version., anyways you will need the contract locally in order to work.

Can I avoid to add every single needed file in my project?

it depends, on the project and the framework you are using. if you main contract just refers to other contract on github you can just import it, because at deploying your will use just your contract. But here we also use `link_token` at deployment, when you do `import LinkToken from brownie` you must have the contract compiled locally on your `build` folder in order to work. So that's why something we copy and paste some contracts and other times we just import it.

Best, regards Cromewar.

**Reply by danmas95** [reply]
Here is some output when I import the necessary files via : `brownie-config.yaml`

1) dependencies: `- smartcontractkit/linktoken` | remappings: `- '@linktoken=smartcontractkit/linktoken'`
OUTPUT: `ValueError: Invalid package ID. Must be given as [ORG]/[REPO]@[VERSION]`

2) dependencies: `- smartcontractkit/linktoken@?.?.?` | remappings: `- '@linktoken=smartcontractkit/linktoken'`
OUTPUT: `ValueError: Invalid version for this package. Available versions are: deployed`

3) dependencies: `- smartcontractkit/linktoken@deployed` | remappings: `- '@linktoken=smartcontractkit/linktoken'`
OUTPUT: 
`ParserError: Source "smartcontractkit/linktoken/contracts/v0.6/ERC677.sol" not found: File not found`
`ParserError: Source "smartcontractkit/linktoken/contracts/v0.6/ITypeAndVersion.sol" not found: File not found.`
`ParserError: Source "smartcontractkit/linktoken/contracts/v0.6/token/LinkERC20.sol" not found: File not found.`

Here is the import I added in `LinkToken.sol` (the v0.6 version you suggested):
![image](https://user-images.githubusercontent.com/50585893/144455411-abe2db41-459d-46e5-a10e-2b220595d6fd.png)

And here is the initial part of the `brownie-config.yaml` file:
![image](https://user-images.githubusercontent.com/50585893/144455631-5c1eaec6-d07f-4339-bb18-5ac282327329.png)

I'm really struggling with this problem and I can't advance in the tutorial due to this...

**Reply by cromewar** [reply]
Yep the correct `@` is deployed, but for some reason is just taking the `v0.7` files and ignoring the other versions, there is some workarounds, choose one of them.

1. first of all upgrade brownie to 1.17 and try using the 0.4 version. 
2. copy all the `v0.6` from github to `.brownie/packages/smartcontractkit/LinkToken@deployed/contracts` so the import would recognize the `v0.6` identifier.
3. use the 0.7 version, but this would require to also copy manually a lot of `sol` files, so I don't recommend it.

Try with option 2, it should work.

**Reply by danmas95** [reply]
I tried to upgrade brownie but...it does not upgrade. 

I used the command `pip install eth-brownie --upgrade` but the brownie version is still `v1.16.4`. Strange.
For every component or package the output is: `Requirement already satisfied: [something]`

I looked for a solution but I couldn't find any. I assume I need to uninstall and re-install brownie, right?

**Reply by cromewar** [reply]
to upgrade pip packages please refer [this](https://monovm.com/blog/how-to-upgrade-pip-package/), I still think point number two would work better for you tough.

**Reply by danmas95** [reply]
I'm trying option number 2 and partially worked.
It recognizes v0.6 directory but some files request the content of `LinkToken/vendor/OpenZeppelin` and this means that I should add one (or more) additional directory to `.brownie/packages/smartcontractkit/LinkToken@deployed/contracts`. 

Here is the complete error displayed: `.brownie/packages/smartcontractkit/LinkToken@deployed/vendor/OpenZeppelin/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol" not found: File not found.`

That's look a bit messy, in my opinion (but seems at least a reasonable solution).

**Reply by cromewar** [reply]
Actually is not a "solution", is more a workaround because I don't know why your brownie refuses to compile the 0.4.11 version, I tried with both brownie 1.16 and 1.17 and I wasn't able to reproduce the problem as mine compiles without problem, may I ask which Os are you using?.

Anyways for this workaround just copy also the `vendor` folder of link token [github](https://github.com/smartcontractkit/LinkToken) on the ` .brownie/packages/smartcontractkit/LinkToken@deployed` path.

**Reply by danmas95** [reply]
I'm on Windows 11, at the moment.

**Reply by cromewar** [reply]
I suggest you to install WSL and manage your python things there, you could refer to the course index, windows support appart. There is a guide about how to configure windows for smart contract development.

**Reply by danmas95** [reply]
I already have WLS installed (I tried it when I had W10 installed). 

Ok, I will give it a try. Anyway, I need to find out what is the problem with brownie. Maybe others following the BC&S course are facing the same problem and finding out a valid solution could help them as well as can solve my problem.

**Reply by cromewar** [reply]
100% if you could find a contundente Solution it'll be very helpful. I don't know why your brownie is refusing to compile, I'm using wsl and 0.4.11 worked fine.

Best regards, Cromewar.

**Reply by danmas95** [reply]
I updated WSL to WSL 2, reinstalled all requirements for BCH&SOL course and now I have this brownie version: `Brownie v1.17.1 - Python development framework for Ethereum`

Compiling on WSL terminal does not give the solidity version error (**at the moment!!**) but now it does not recognize anymore the OpenZeppelin repository where Ownable is located...
The error is: `FileNotFoundError: [Errno 2] No such file or directory: '[PATH].brownie/packages/OpenZeppelin/openzeppelin-contracts@3.4.0/contracts/access/Ownable.sol'`

**Reply by cromewar** [reply]
That's strange as `.brownie` folder is different for windows and Linux, so when you compile using WSL it should create that folder on the linux path, so if I have to guess I would say you are using vscode but using windows. You would have to install the WSL extension and run a vscode windows on WSL, also is highly recommended to move your project folder to WSL home path.

If you want a guide about how to configure it you may refer to this [link](https://medium.com/@cromewar/how-to-setup-windows-10-11-for-smart-contract-development-and-brownie-e7d8d13555b3).

**Reply by danmas95** [reply]
Yes, I wrote the code on VSCode.

Ok, I will try what you suggested. Thanks!

**Reply by danmas95** [reply]
I set everything up but stil...not working.

VSCode il linked to WSL and I can access and modify file from VSC and Ubuntu Terminal on Windows. The HUGE problem is that the the file `OpenZeppelin/openzeppelin-contracts@3.4.0/contracts/access/Ownable.sol` seems to be not recognized. 
At first I thought that the problem was that file/folder was stored in `C:/[SOME_PATH]/.brownie/packages` but when I tried to copy it into the equivalent `brownie` folder on WSL...well...the `OpenZeppelin` folder was still not recognized as file or diretory (`No such file or directory`).

As usual: `FileNotFoundError: [Errno 2] No such file or directory: '.brownie/packages/OpenZeppelin/openzeppelin-contracts@3.4.0/contracts/access/Ownable.sol'`

This situation is becoming frustrating...

**Reply by cromewar** [reply]
That's strange because brownie is not recognizing the packages, they should be on this path:

![image](https://user-images.githubusercontent.com/20199180/144728228-50c76668-8b61-4da4-8aee-230c0c6853ad.png)

Just as a test, could you clone the lesson [repo](https://github.com/PatrickAlphaC/smartcontract-lottery) into linux path:

![image](https://user-images.githubusercontent.com/20199180/144728371-f2f6593f-3695-438b-b74a-6fb34f0c4d14.png)

and try to compile the project using brownie compile.

This is also frustrating for me as I wasn't able to reproduce the error, using same WSL mine compiles the same code without problems.

Note: also is recommended to use a new python virtualenv.

**Reply by cromewar** [reply]
@PatrickAlphaC could you give us a tip on this one?

**Reply by danmas95** [reply]
I manually copied OpenZeppelin and Smartcontractkit folders into .brownie in WSL and didn't recognized it. It still looks for the folder in the Windows path. Strage, very strange.

**Reply by cromewar** [reply]
Actually those files will appear automatically after doing a compile installing into brownie folder, so coping directly to that path wont work, that's why I suggested you to clone the repo and compile, because those files should be generated by brownie, so I recommend you to delete them and compile a project using brownie on WSL.

**Reply by danmas95** [reply]
[UPDATE] - Current situation

I deleted the project on WSL and started a new one, as you suggested. I copied every single file from "my version" of the project (I watched the video course and coded along with Pat, writing here and there a couple of notes, so I prefered to reuse my code).

I compiled via `brownie compile` and **NO ERRORS** came out!! Incredible!

Just a couple of warnings came out, to be honest. In particular:

1) `WARNING: Unable to compile smartcontractkit/chainlink-brownie-contracts@1.1.1 due to a NamespaceCollision - you may still be able to import sources from the package, but will be unable to load the package directly.`

2) `WARNING: Unable to compile OpenZeppelin/openzeppelin-contracts@3.4.0 due to a CompilerError - you may still be able to import sources from the package, but will be unable to load the package directly.`

Anyway...it seems everything fine (at the moment, at least...).

So...**THANK YOU!** You really helped me.


Now...I would like to know why the problem came out on Windows and how can be solved...but for now, I'm happy with the "WSL Solution".

**Reply by cromewar** [reply]
WARNING: Unable to compile smartcontractkit/chainlink-brownie-contracts@1.1.1 this just means you have two files which are the exact same, brownie is saying you have the local file on your project and the one which comes from `Github`, but you don't have to worry about.

.I would like to know why the problem came out on Windows and how can be solve

I have no idea, I a lazy answer is that windows manages packages very poorly and it's a mess managing python stuff, but a concrete answer is something out of my limited knowledge-
 
I'm glad the problem is now solved for you, I'll close this thread now. Feel free to make any other questions you want.
